**Import packages**

In [1]:
import numpy as np
import subprocess
from Bio import Entrez
from Bio import SeqIO

## Simulating the data
Write sample data files containing 100%, 90%, 80%, 70%, 60%, and 50% zeros (100Mb each) automately.     
100 * (1024kb/mb) * (1024bytes/kb) * (8bits/byte)      

In [2]:
p_zero = [["zeros_100p", [1, 0]], 
          ["zeros_90p", [0.9, 0.1]], 
          ["zeros_80p", [0.8, 0.2]],
          ["zeros_70p", [0.7, 0.3]],
          ["zeros_60p", [0.6, 0.4]],
          ["zeros_50p", [0.5, 0.5]]
         ]


**np.random.choice**     
first arg: the sample pool   
size: sample size     
replace: whether or not the sample is returned to sample pool    
p: probability    

In [3]:
# Run this once
for i in p_zero:
    myvar = np.random.choice([0, 1], size=(8, 1024, 1024, 100), replace=True, p=i[1])
    myvar = np.packbits(myvar)
    open(i[0], "wb").write(myvar)

#### Generate 100-million-letter-long DNA sequences 


In [15]:
# Run this once
my_nt_seq = np.random.choice(['A', 'T', 'C', 'G'], size=100000000, p=[0.25, 0.25, 0.25, 0.25])
open('nt_seq.fa', 'w').write(''.join(my_nt_seq))

100000000

#### Generate 100-million-letter-long protein sequences 

In [17]:
# Run this once
my_protein_seq = np.random.choice(['R', 'K', 'D', 'E', 
                      'Q', 'N', 'H', 'S', 'T', 'Y', 'C', 'W',
                      'A', 'I', 'L', 'M', 'F', 'V', 'P', 'G'],
                     size=(100000000), p=[1/20 for n in range(20)])
open('protein_seq.fa', 'w').write(''.join(my_protein_seq))

100000000

## Terminal

Here are the commands I will be using in terminal.I will run the codes in separate trunks.

In [3]:
cmds_gzip = []
cmds_bzip2 = []
cmds_pbzip2 = []
cmds_art = []

for i in p_zero:
    cmds_gzip.append('time gzip -k ' + i[0])
    cmds_bzip2.append('time bzip2 -k '+ i[0] )
    cmds_pbzip2.append('time pbzip2 -k ' + i[0])
    cmds_art.append('time ArithmeticCompress ' + i[0] + ' ' + i[0] + '.art')
    
cmds_gzip.append('time gzip -k nt_seq.fa')
cmds_bzip2.append('time bzip2 -k nt_seq.fa')
cmds_pbzip2.append('time pbzip2 -k nt_seq.fa')
cmds_art.append('time ArithmeticCompress nt_seq.fa  nt_seq.art')

cmds_gzip.append('time gzip -k protein_seq.fa')
cmds_bzip2.append('time bzip2 -k protein_seq.fa')
cmds_pbzip2.append('time pbzip2 -k protein_seq.fa')
cmds_art.append('time ArithmeticCompress protein_seq.fa  protein_seq.art')

**Run gzip**

In [7]:
# Run once
for cmd in cmds_gzip:
    print(cmd)
    !{cmd}
    print()

time gzip -k zeros_100p
0.85user 0.01system 0:00.87elapsed 99%CPU (0avgtext+0avgdata 1908maxresident)k
0inputs+200outputs (0major+146minor)pagefaults 0swaps

time gzip -k zeros_90p
18.82user 0.09system 0:18.92elapsed 99%CPU (0avgtext+0avgdata 1816maxresident)k
0inputs+114720outputs (0major+139minor)pagefaults 0swaps

time gzip -k zeros_80p
13.35user 0.11system 0:13.46elapsed 99%CPU (0avgtext+0avgdata 1776maxresident)k
0inputs+158520outputs (0major+131minor)pagefaults 0swaps

time gzip -k zeros_70p
7.45user 0.19system 0:07.65elapsed 99%CPU (0avgtext+0avgdata 1764maxresident)k
0inputs+182848outputs (0major+130minor)pagefaults 0swaps

time gzip -k zeros_60p
4.97user 0.18system 0:05.16elapsed 100%CPU (0avgtext+0avgdata 1788maxresident)k
0inputs+200040outputs (0major+204minor)pagefaults 0swaps

time gzip -k zeros_50p
4.28user 0.18system 0:04.47elapsed 100%CPU (0avgtext+0avgdata 1800maxresident)k
0inputs+204840outputs (0major+132minor)pagefaults 0swaps

time gzip -k nt_seq.fa
12.14user 0.07s

**Run bzip2**

In [8]:
# Run once
for cmd in cmds_bzip2:
    print(cmd)
    !{cmd}
    print()

time bzip2 -k zeros_100p
1.03user 0.05system 0:01.08elapsed 100%CPU (0avgtext+0avgdata 8828maxresident)k
0inputs+8outputs (0major+1904minor)pagefaults 0swaps

time bzip2 -k zeros_90p
13.37user 0.12system 0:13.50elapsed 99%CPU (0avgtext+0avgdata 7724maxresident)k
0inputs+119448outputs (0major+1690minor)pagefaults 0swaps

time bzip2 -k zeros_80p
11.93user 0.13system 0:12.07elapsed 99%CPU (0avgtext+0avgdata 7716maxresident)k
0inputs+169232outputs (0major+1691minor)pagefaults 0swaps

time bzip2 -k zeros_70p
13.82user 0.14system 0:13.96elapsed 99%CPU (0avgtext+0avgdata 7824maxresident)k
0inputs+194848outputs (0major+1691minor)pagefaults 0swaps

time bzip2 -k zeros_60p
15.68user 0.16system 0:15.85elapsed 99%CPU (0avgtext+0avgdata 7720maxresident)k
0inputs+204904outputs (0major+1690minor)pagefaults 0swaps

time bzip2 -k zeros_50p
20.90user 0.17system 0:21.08elapsed 99%CPU (0avgtext+0avgdata 7732maxresident)k
0inputs+205720outputs (0major+1690minor)pagefaults 0swaps

time bzip2 -k nt_seq.fa
11

**Run pbzip2**

In [11]:
# Run once
for cmd in cmds_pbzip2:
    print(cmd)
    !{cmd}
    print()

time pbzip2 -k zeros_100p
1.94user 0.12system 0:00.10elapsed 1993%CPU (0avgtext+0avgdata 48688maxresident)k
0inputs+16outputs (0major+25958minor)pagefaults 0swaps

time pbzip2 -k zeros_90p
19.32user 0.80system 0:00.77elapsed 2590%CPU (0avgtext+0avgdata 275644maxresident)k
0inputs+119504outputs (0major+223861minor)pagefaults 0swaps

time pbzip2 -k zeros_80p
23.36user 0.89system 0:00.93elapsed 2605%CPU (0avgtext+0avgdata 281984maxresident)k
0inputs+169264outputs (0major+228748minor)pagefaults 0swaps

time pbzip2 -k zeros_70p
29.44user 0.85system 0:01.13elapsed 2664%CPU (0avgtext+0avgdata 285340maxresident)k
0inputs+194864outputs (0major+243610minor)pagefaults 0swaps

time pbzip2 -k zeros_60p
36.15user 0.88system 0:01.38elapsed 2666%CPU (0avgtext+0avgdata 288592maxresident)k
0inputs+204920outputs (0major+259600minor)pagefaults 0swaps

time pbzip2 -k zeros_50p
39.97user 0.90system 0:01.51elapsed 2690%CPU (0avgtext+0avgdata 288816maxresident)k
0inputs+205736outputs (0major+271634minor)pagef

**Run ArithmetricCompress**

In [12]:
# Run once
for cmd in cmds_art:
    print(cmd)
    !{cmd}
    print()

time ArithmeticCompress zeros_100p zeros_100p.art
17.60user 0.09system 0:17.70elapsed 99%CPU (0avgtext+0avgdata 4344maxresident)k
0inputs+8outputs (0major+231minor)pagefaults 0swaps

time ArithmeticCompress zeros_90p zeros_90p.art
28.65user 0.20system 0:28.86elapsed 99%CPU (0avgtext+0avgdata 4360maxresident)k
0inputs+96056outputs (0major+234minor)pagefaults 0swaps

time ArithmeticCompress zeros_80p zeros_80p.art
35.42user 0.27system 0:35.69elapsed 99%CPU (0avgtext+0avgdata 4212maxresident)k
0inputs+147856outputs (0major+232minor)pagefaults 0swaps

time ArithmeticCompress zeros_70p zeros_70p.art
39.17user 0.38system 0:39.56elapsed 99%CPU (0avgtext+0avgdata 4164maxresident)k
0inputs+180496outputs (0major+233minor)pagefaults 0swaps

time ArithmeticCompress zeros_60p zeros_60p.art
51.66user 0.45system 0:52.11elapsed 99%CPU (0avgtext+0avgdata 4284maxresident)k
0inputs+198856outputs (0major+232minor)pagefaults 0swaps

time ArithmeticCompress zeros_50p zeros_50p.art
51.40user 0.41system 0:51.

|command |input file | input size| output file | output size| time|
|--------|-----------|-----------|-------------|------------|-----|
|**gzip**|zeros_100p|105 MB|zeros_100p.gz|102 KB|0.85s|
|gzip|zeros_90p|105 MB|zeros_90p.gz|58.7 MB|18.82s|
|gzip|zeros_80p|105 MB|zeros_80p.gz|81.2 MB|13.35s|
|gzip|zeros_70p|105 MB|zeros_70p.gz|93.6 MB|7.45s|
|gzip|zeros_60P|105 MB|zeros_60p.gz|102 MB|4.97s|
|gzip|zeros_50p|105 MB|zeros_50p.gz|105 MB|4.28s|
|gzip|nt_seq.fa|100 MB|nt_seq.fa.gz|29.2 MB|12.14s|
|gzip|protein_seq.fa|100 MB|protein_seq.fa.gz|60.6 MB|4.46s|
|--------|-----------|-----------|-------------|------------|-----|
|**bzip2**|zeros_100p|105 MB|zeros_100p.bz2|113 B|1.03s|
|bzip2|zeros_90p|105 MB|zeros_90p.bz2|61.2 MB|13.37s|
|bzip2|zeros_80p|105 MB|zeros_80p.bz2|86.6 MB|11.93s|
|bzip2|zeros_70p|105 MB|zeros_70p.bz2|99.8 MB|13.82s|
|bzip2|zeros_60P|105 MB|zeros_60p.bz2|105 MB|15.68s|
|bzip2|zeros_50p|105 MB|zeros_50p.bz2|105 MB|20.90s|
|bzip2|nt_seq.fa|100 MB|nt_seq.fa.bz2|27.3 MB|11.95s|
|bzip2|protein_seq.fa|100 MB|protein_seq.fa.bz2|55.3 MB|12.53s|
|--------|-----------|-----------|-------------|------------|-----|
|**pbzip2**|zeros_100p|105 MB|zeros_100p.bz2|5.62 KB|1.94s|
|pbzip2|zeros_90p|105 MB|zeros_90p.bz2|61.2 MB|19.32s|
|pbzip2|zeros_80p|105 MB|zeros_80p.bz2|86.7 MB|23.36s|
|pbzip2|zeros_70p|105 MB|zeros_70p.bz2|99.8 MB|29.44s|
|pbzip2|zeros_60P|105 MB|zeros_60p.bz2|105 MB|36.15s|
|pbzip2|zeros_50p|105 MB|zeros_50p.bz2|105 MB|39.97s|
|pbzip2|nt_seq.fa|100 MB|nt_seq.fa.bz2|27.3 MB|15.91s|
|pbzip2|protein_seq.fa|100 MB|protein_seq.fa.bz2|55.3 MB|18.37s|
|--------|-----------|-----------|-------------|------------|-----|
|**ArithmeticCompress**|zeros_100p|105 MB|zeros_100p.bz2|1.03 KB|17.60s|
|ArithmeticCompress|zeros_90p|105 MB|zeros_90p.art|49.2 MB|28.65s|
|ArithmeticCompress|zeros_80p|105 MB|zeros_80p.art|75.7 MB|35.42s|
|ArithmeticCompress|zeros_70p|105 MB|zeros_70p.art|92.4 MB|39.17s|
|ArithmeticCompress|zeros_60P|105 MB|zeros_60p.art|102 MB|51.66s|
|ArithmeticCompress|zeros_50p|105 MB|zeros_50p.art|105 MB|51.40s|
|ArithmeticCompress|nt_seq.fa|100 MB|nt_seq.fa.art|25 MB|26.94s|
|ArithmeticCompress|protein_seq.fa|100 MB|protein_seq.fa.art|54 MB|28.27s|

**Which algorithm achieves the best level of compression on each file type?**    
ArithmeticCompress achieves the best level of compression on both .wb and .fa files. 

**Which algorithm is the fastest?**   
bzip2 is the fastest for all files except for zeros_100p.  

**What is the difference between bzip2 and pbzip2? Do you expect one to be faster and why?**     
bzip2 compresses data in blocks of size between 100~900kB. Its algorithm converts frequently-recurring character sequences into strings of identitital letters. bzip2 performance is asymmetric, as decompression is relatively fast.    
pbzip2, a modified version of bzip2, supports multi-threading (execute threads parallelly), giving almost linear speed improvements on multi-CPU and multi-core computers. (Note: info extracted from wikipedia)      
I expect pbzip2 to be faster because it runs multiple threads at the same time whereas bzip2 only runs single thread.     

**How does the level of compression change as the percentage of zeros increases? Why does this happen?**    
When zeros increases, level of compression increases.    

**What is the minimum number of bits required to store a single DNA base?**    
2 bits per DNA base.

**What is the minimum number of bits required to store an amino acid letter?**    
4.322 bits (log20) per amino acid.

**In your tests, how many bits did gzip and bzip2 actually require to store your random DNA and protein sequences?**    
*DNA gzip:*     
DNA seq has 100 million nucleotides and size 100MB (8bits/nt).     
gzip compresses it to 29.2MB (29.2MB * 1024KB/MB * 1024B/KB * 8bits/B = 244947353.6 bytes).    
244947353.6 bytes/100 million nts = 2.449 bytes/nucleotide.    

*DNA bzip2:*     
DNA seq has 100 million nucleotides and size 100MB (8bits/nt).     
bzip2 compresses it to 27.3MB (27.3MB * 1024KB/MB * 1024B/KB * 8bits/B = 229008998.4 bytes).    
229008998.4 bytes/100 million nts = 2.290 bytes/nucleotide. 

*Protein gzip:*     
Protein seq has 100 million AA and size 100MB (8bits/aa).     
gzip compresses it to 60.6MB (60.6MB * 1024KB/MB * 1024B/KB * 8bits/B = 508349644.8 bytes).    
508349644.8 bytes/100 million aas = 5.08 bytes/AA.   

*Protein bzip2:*     
Protein seq has 100 million AA and size 100MB (8bits/aa).     
gzip compresses it to 55.3MB (55.3MB * 1024KB/MB * 1024B/KB * 8bits/B = 463890022.4 bytes).    
463890022.4 bytes/100 million aas = 4.639 bytes/AA.

**Are gzip and bzip2 performing well on DNA and proteins?**    
Yes. gzip and bzip2 compress DNA files into 29.2MB and 27.3MB and protein files into 55.3MB and 55.3MB respectively. bzip2 takes less bytes to store each nucleotide and amino acid.    

## Compressing real data
find the nucleic acid sequences of gp120 homologs from at least 10 different HIV isolates and concatenate them together into a single multi-FASTA.

### Entrez 

Search the nuclueic acid sequences of gp120 homologs from 10 different HIV isolates in Entrez nucleotide database. Get the IDs.

**A priori, do you expect to achieve better or worse compression here than random data? Why?**     
I expect to achieve better compression here than random data because the fasta file here contains shorter sequences, which requires less specificity.


In [37]:
# Run once
Entrez.email = "jinghuawu@berkeley.edu"
handle = Entrez.esearch(db='nucleotide', 
                        term='gp120 hiv', 
                        sort='relevance',
                        idtype='acc',
                        retmax=10)
records = Entrez.read(handle)
print (records['IdList'])

['MG515154.1', 'MG515153.1', 'MG515152.1', 'MG515151.1', 'MG515150.1', 'MG515149.1', 'MG515148.1', 'MG515147.1', 'MG515146.1', 'MG515145.1']


Use the result IDs to fetch info from nucleotide database, rettype is fasta. Use SeqIO to read the fasta result and write a fasta file containing multiple fasta results.

In [38]:
# Run once
ids = records['IdList']
f = open('gp120_HIVisolates.fa', 'w')
for i in ids:
    handle = Entrez.efetch(db='nucleotide', 
                          id=i, 
                          rettype='fasta', 
                          retmode='text')
    record = SeqIO.read(handle, 'fasta')
    fasta = SeqIO.write(record, f, "fasta")
f.close()

## Compress the multi-FASTA using gzip, bzip2, and arithmetic coding

### Terminal

In [42]:
# Run once
! time gzip -k gp120_HIVisolates.fa
! time bzip2 -k gp120_HIVisolates.fa
! mv gp120_HIVisolates.fa.bz2 gp120_HIVisolates_bzip2.fa.bz2 # change name to prevent overwrite
! time pbzip2 -k gp120_HIVisolates.fa
! time ArithmeticCompress gp120_HIVisolates.fa gp120_HIVisolates.art

0.00user 0.00system 0:00.00elapsed 100%CPU (0avgtext+0avgdata 1640maxresident)k
0inputs+8outputs (0major+94minor)pagefaults 0swaps
bzip2: Output file gp120_HIVisolates.fa.bz2 already exists.
Command exited with non-zero status 1
0.00user 0.00system 0:00.00elapsed 100%CPU (0avgtext+0avgdata 1604maxresident)k
0inputs+0outputs (0major+70minor)pagefaults 0swaps
0.00user 0.00system 0:00.00elapsed 100%CPU (0avgtext+0avgdata 3180maxresident)k
0inputs+8outputs (0major+152minor)pagefaults 0swaps
0.00user 0.00system 0:00.00elapsed 100%CPU (0avgtext+0avgdata 4232maxresident)k
0inputs+8outputs (0major+234minor)pagefaults 0swaps


**How does the compression ratio of this file compare to random data?**   
Except for ArithmeticCompress, all other algorithms generate higher compression ratios in the gp120 fasta data compare to the random data (see table below).

|command |input file | input size| output file | output size| time|ratio(in/out)| 
|--------|-----------|-----------|-------------|------------|-----|-----|
|gzip|nt_seq.fa|100 MB|nt_seq.fa.gz|29.2 MB|12.14s|3.42|
|gzip|gp120_HIVisolates.fa|6.05KB|gp120_HIVisolates.fa.gz|821B|0.00s|7.546|
|bzip2|nt_seq.fa|100 MB|nt_seq.fa.bz2|27.3 MB|11.95s|3.66|
|bzip2|gp120_HIVisolates.fa|6.05KB|gp120_HIVisolates.fa.bz2|998B|0.00s|6.208|
|pbzip2|nt_seq.fa|100 MB|nt_seq.fa.bz2|27.3 MB|15.91s|3.66|
|pbzip2|gp120_HIVisolates.fa|6.05KB|gp120_HIVisolates.fa.bz2|998B|0.00s|6.208|
|ArithmeticCompress|nt_seq.fa|100 MB|nt_seq.fa.art|25 MB|26.94s|4|
|ArithmeticCompress|gp120_HIVisolates.fa|6.05KB|gp120_HIVisolates.fa.art|3.37KB|0.00s|1.795|

## Estimate compression of 1000 terabytes

**Given the benchmarking data you obtained in this lab, which algorithm do you propose to use for each type of data? Provide an estimate for the fraction of space you can save using your compression scheme. How much of a bonus do you anticipate receiving this year?**

For the 80% re-sequencing of genomes and plasmids that are very similar to each other: 
Use **gzip**, the one uses DEFLATE (Huffman + Lempel-Ziv algo). As shown in the previous table, gzip performs the best job on compressing nucleotide datas that are similar to each other. There are so many mutual information between these similar genomes. Therefore gzip could be the best one here.     
The ratio of before-after-gzip is 7.546 for gp120_HIVisolates.fa.    
1000tb x (1tb/1024gb) x (1024mb/1gb) x (1024kb/1mb) x (1024b/1kb) x (1024bytes/1b) is about 8e+12 bits.        
0.8 x (8 x 10 ^ 12) = 6.4 x 10 ^ 12           
6.4 x (10 ^ 12) / 7.546 = 848131460376.3583 bits total for nucleotides     

For the 10% completely random protein sequences:     
Use **ArithmeticCompress**. The first table shows ArithmeticCompress compresses the randomly generated protein sequence the best (from 100MB to 54MB, which is 2:1). Therefore:     
0.1 x (8 x 10 ^ 12) = 8 x 10 ^ 11 
8 x (10 ^ 11) / 2 = 400000000000.0 bits total for aa 

For the last 10% binary images:     
Use **bzip2**, the one that uses run-length coding. Run-length encoding compresses runs of identical pixels. Since the images are binary, we can use the 0-1 data sample previously generated. The worse case for binary image here is half 0 and 1. Bzip2 did not compress the data but saved time. Therefore:       
0.1 x (8 x 10 ^ 12) = 800000000000 bits total for binary images     

Total: 848131460376.3583 + 400000000000.0 + 800000000000 = 2048131460376.3584 bits       

%saved = (before - after) / before
((6.4 x (10 ^ 12)) - 2048131460376.3584) /  (6.4 x (10 ^ 12)) = 0.680 = 68%

$50 x 68 = **3400-dollar bonus**  




    
